Add validation webhook for service profiles

alpeb opened the pull request:
Fixes #2075

The new webhook endpoint is created in a new container in the
linkerd-controller pod.

The proxy injector webhook was refactored to extract-out the scaffolding
to be reused by the validation webhook (and eventually by any other
admission webhooks we'd like to create in the future). The new common
stuff now reside under `pkg/webhook`.

The validation webhook simply sets up the endpoint server and the
webhook config using `pkg/webhook`, and calls the old
`profiles.Validate` to perform the actual validation.

I Also got rid of the unused `-no-init-container` and `-tlsEnabled`
flags in the proxy injector.

Todo in this PR: tests.

Todo in a follow-up PRs: remove the SP check from the CLI check.

grampelberg commented:
@alpeb cool! Looks like it does a pretty good job explaining what broke as well =)

alpeb commented:
@adleong the dependencies changes were caused by having mistakingly imported Prometheus's logger instead of the usual one. I fixed that along with tests.

siggy commented on controller/sp-validator/webhook.go:
```diff
@@ -0,0 +1,22 @@
+	admissionResponse := &admissionv1beta1.AdmissionResponse{Allowed: true}
+	if err := profiles.Validate(request.Object.Raw); err != nil {
+		admissionResponse.Allowed = false
+		admissionResponse.Result = &metav1.Status{Message: err.Error()}
```
nice!

alpeb commented on pkg/webhook/server.go:
```diff
@@ -0,0 +1,154 @@
+package webhook
```
You're right, I'll move all that over `controller/webhook`.

alpeb commented on controller/cmd/sp-validator/main.go:
```diff
@@ -0,0 +1,69 @@
+package main
```
I've refactored both `main.go` s so they just set up the config and use a new common `launcher.go`.
Both `webhook_ops.go` have a similar structure but use different types all over, so I think it's better to leave them both. Also it seems to me both template files are different enough to leave them as separate files.
As for the tests, `webhook_test.go` verifies that the auto injection works as expected.
The service profile validator is really just a wrapper around `profiles.Validate()` which already has unit tests inside its package. The wrapping part is also already tested a bit in `server_test.go``

olix0r replied:
It's worth noting that much of the bootstrapping stuff is intended to be pretty-short lived. In 2.4, we plan to move webhook creation into the installation process.

alpeb commented on controller/proxy-injector/webhook.go:
```diff
@@ -97,7 +33,7 @@ func (w *Webhook) inject(request *admissionv1beta1.AdmissionRequest) (*admission
 	}
 
-	namespace, err := w.client.CoreV1().Namespaces().Get(request.Namespace, metav1.GetOptions{})
+	namespace, err := client.CoreV1().Namespaces().Get(request.Namespace, metav1.GetOptions{})
```
Good point :+1:  I've issued #2566 to track that work.

alpeb commented:
I've addressed all the feedback in my latest push. But thinking on rebasing after #2555, #2547 and #2500 land because this is gonna be painful, so no need to endure that pain in those branches as well :hurtrealbad:

alpeb commented:
Superseded by #2623
